### PR TITLE
ZipArchive now disposes given stream only once

### LINKED ISSUE #79695: ZipArchive disposes passed stream twice
### Description

I've tracked down a bug in our software where we had a stream which was not safe to call dispose multiple times (Fixed in the meantime =)). 
This surfaced when we passed said stream into a ZipArchive (reading only), because it is disposing the passed stream twice. 
I know that this is technically not a bug, but nonetheless something which can be improved.

The issue itself is caused by [this line](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L167). The BinaryReader gets passed the stream, but the leaveOpen parameter is not set. Hence when [disposing](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L464-L466) it disposes the stream twice.

### Reproduction Steps

- Create a erronous Stream implementation that has an issue when disposing twice.
- Create a ZipArchive in reading mode and dispose it

### Expected behavior

While not technically breaking, it still would be nice not to dispose things multiple times if not needed.

### Actual behavior

Disposes the input stream twice, once directly and once due to the BinaryReader.

### Regression?

not that I'm aware of

### Known Workarounds

Fix the wrongfully implemented stream :) 

### Configuration

I've tested it on .NET6, but it still exists on current main

### Other information

_No response_

### PR DESCRIPTION
**Pull request by etemi**
Fixes #79695

Even though the `IDisposable` contract says that you should be able to call `Dispose` multiple times, be forgiving for implementations that throw when `Dispose` is called multiple times.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
cc @stephentoub as I think you were involved in similar discussions in the past

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs [hunk 1]
```diff
@@ -249,5 +249,37 @@ public static async Task IdentifyEncryptedEntries(string zipFile)
...
+        private class DisposeCallCountingStream : MemoryStream
+        {
+            public int NumberOfDisposeCalls { get; private set; }
+
+            protected override void Dispose(bool disposing)
+            {
+                NumberOfDisposeCalls++;
+                base.Dispose(disposing);
+            }
+        }
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs:283]: Do we have tests that validate the underlying Stream is closed when the ZipArchive is disposed according to leaveOpen?

Comment (etemi) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs:285]: I have augmented the test to be a theory. It passes the leaveOpen parameter now explicitly and tests for the expected outcome. Thanks for your suggestion @stephentoub

**Review by stephentoub — APPROVED** [review]
Thanks

**Pull request merged**